<a href="https://colab.research.google.com/github/abhi123412/IPSVIP/blob/main/FakeNewsDetection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Machine Learning Intern at InfoPillar Solutions Pvt Ltd - Virtual Internship Program**

# Author - Abhijeet Balu Thombare

# **Task** 1 :- Fake News Detection Project
Project idea :– Fake news spreads like a wildfire and this is a big issue in this era

#1] Import Libraries

In [41]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# 2] Load Dataset

In [42]:
#Read the data
dataset = pd.read_csv('/content/news.csv')

#3] Summarise Dataset

In [43]:
print("No. of Columns and Rows :- ",dataset.shape) #It gives no. of rows and columns

print(dataset.head()) #It gives first five record

No. of Columns and Rows :-  (6335, 4)
   Unnamed: 0  ... label
0        8476  ...  FAKE
1       10294  ...  FAKE
2        3608  ...  REAL
3       10142  ...  FAKE
4         875  ...  REAL

[5 rows x 4 columns]


#4] Get the Labels from DataSet

In [44]:
labels=dataset.label
# labels.head(10)
print(labels)

0       FAKE
1       FAKE
2       REAL
3       FAKE
4       REAL
        ... 
6330    REAL
6331    FAKE
6332    FAKE
6333    REAL
6334    REAL
Name: label, Length: 6335, dtype: object


#5] Split the dataset into training and testing sets

In [45]:
#Spliting the dataset into testing and training
x_train,x_test,y_train,y_test =train_test_split(dataset['text'], labels, test_size=0.2, random_state=7)

#6] Fit and transform the vectorizer on the train set, and transform the vectorizer on the test set.

In [47]:
#Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

In [48]:
#Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

#7] Initialize a PassiveAggressiveClassifier.

In [49]:
#Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

In [50]:
#Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.58%


#The accuracy of 92.58% with this model.

#8]Print out a confusion matrix.

In [51]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[588,  50],
       [ 44, 585]])

#In this model, we have 588 true positives, 585 true negatives, 50 false 
#positives, and 44 false negatives.
